From 83e3d4b0ae373dcba30c68bf28f8d179191a297a Mon Sep 17 00:00:00 2001
From: valis <sec@valis.email>
Date: Sat, 29 Jul 2023 08:32:01 -0400
Subject: net/sched: cls_fw: No longer copy tcf_result on update to avoid
 use-after-free

[ Upstream commit 76e42ae831991c828cffa8c37736ebfb831ad5ec ]

When fw_change() is called on an existing filter, the whole
tcf_result struct is always copied into the new instance of the filter.

This causes a problem when updating a filter bound to a class,
as tcf_unbind_filter() is always called on the old instance in the
success path, decreasing filter_cnt of the still referenced class
and allowing it to be deleted, leading to a use-after-free.

Fix this by no longer copying the tcf_result struct from the old filter.

Fixes: e35a8ee5993b ("net: sched: fw use RCU")
Reported-by: valis <sec@valis.email>
Reported-by: Bing-Jhong Billy Jheng <billy@starlabs.sg>
Signed-off-by: valis <sec@valis.email>
Signed-off-by: Jamal Hadi Salim <jhs@mojatatu.com>
Reviewed-by: Victor Nogueira <victor@mojatatu.com>
Reviewed-by: Pedro Tammela <pctammela@mojatatu.com>
Reviewed-by: M A Ramdhan <ramdhan@starlabs.sg>
Link: https://lore.kernel.org/r/20230729123202.72406-3-jhs@mojatatu.com
Signed-off-by: Jakub Kicinski <kuba@kernel.org>
Signed-off-by: Sasha Levin <sashal@kernel.org>
[Srish: Fixes CVE-2023-4128. The CVE is fixed by a total of three commits.
        Out of the three, two have been merged to the 4.19.x stable tree.
        These are the merged commits.
            - stable commit id: 4aae24015ecd70d824a953e2dc5b0ca2c4769243
            - stable commit id: ad8f36f96696a7f1d191da66637c415959bab6d8
        This patch addresses the missing commit
            - upstream commit id: 76e42ae831991c828cffa8c37736ebfb831ad5ec]
Signed-off-by: Srish Srinivasan <ssrish@vmware.com>
---
 net/sched/cls_fw.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/net/sched/cls_fw.c b/net/sched/cls_fw.c
index 5284a473c..f15089c24 100644
--- a/net/sched/cls_fw.c
+++ b/net/sched/cls_fw.c
@@ -277,7 +277,6 @@ static int fw_change(struct net *net, struct sk_buff *in_skb,
 			return -ENOBUFS;
 
 		fnew->id = f->id;
-		fnew->res = f->res;
 #ifdef CONFIG_NET_CLS_IND
 		fnew->ifindex = f->ifindex;
 #endif /* CONFIG_NET_CLS_IND */
-- 
2.35.6

